File: Sources/FaceColour/ColorEngine/Season.swift
import Foundation

/// 4-season color type (MVP). Designed to extend to a 12-season model later.
enum Season: String, Codable, Equatable, CaseIterable {
    case spring, summer, autumn, winter

    var displayName: String { rawValue.capitalized }

    /// Classify from undertone + depth (Phase 3 MVP — contrast not yet used).
    ///
    /// - Warm + light → Spring, Warm + deep → Autumn
    /// - Cool + light → Summer, Cool + deep → Winter
    /// - Neutral leans warm/cool by the Lab hue angle.
    static func classify(undertone: Undertone, depth: Fitzpatrick, hueAngle: Double) -> Season {
        let warmLeaning: Bool
        switch undertone {
        case .warm: warmLeaning = true
        case .cool: warmLeaning = false
        case .neutral: warmLeaning = hueAngle >= SeasonThresholds.neutralWarmHue
        }

        let deep = SeasonThresholds.deepDepths.contains(depth)
        switch (warmLeaning, deep) {
        case (true, false): return .spring
        case (true, true): return .autumn
        case (false, false): return .summer
        case (false, true): return .winter
        }
    }
}

enum SeasonThresholds {
    /// Neutral undertones with a hue angle at/above this lean warm.
    static let neutralWarmHue = 51.0
    /// Fitzpatrick phototypes considered "deep" for the season split (IV–VI).
    static let deepDepths: Set<Fitzpatrick> = [.typeIV, .typeV, .typeVI]
}

File: android/app/src/main/java/com/facecolour/app/engine/Season.kt
package com.facecolour.app.engine

/** 4-season color type (MVP), extensible to 12-season later. */
enum class Season {
    SPRING, SUMMER, AUTUMN, WINTER;

    val displayName: String get() = name.lowercase().replaceFirstChar { it.uppercase() }

    companion object {
        /** From undertone + Fitzpatrick depth; neutral leans by hue angle. */
        fun classify(undertone: Undertone, fitzpatrick: Fitzpatrick, hueAngle: Double): Season {
            val warm = when (undertone) {
                Undertone.WARM -> true
                Undertone.COOL -> false
                Undertone.NEUTRAL -> hueAngle >= SeasonThresholds.NEUTRAL_WARM_HUE
            }
            val deep = fitzpatrick in SeasonThresholds.DEEP_TYPES
            return when {
                warm && !deep -> SPRING
                warm && deep -> AUTUMN
                !warm && !deep -> SUMMER
                else -> WINTER
            }
        }
    }
}

object SeasonThresholds {
    const val NEUTRAL_WARM_HUE = 51.0
    val DEEP_TYPES = setOf(Fitzpatrick.TYPE_IV, Fitzpatrick.TYPE_V, Fitzpatrick.TYPE_VI)
}
